## Introduction

![](https://download.logo.wine/logo/ASOS_(retailer)/ASOS_(retailer)-Logo.wine.png)

It's your first day on the job and you're given data about the weekly revenue at ASOS.

In the list named `revenue_by_week`, you'll find a snippet of ASOS's estimated weekly revenue, captured in millons of pounds.

But, ugh! There seems to be some bugs with how the data was stored in the list 😭

In [53]:
revenue_by_week = [65, 77, '66', '74', 
                   64, 82, '86', 72, '80',
                   96, 101, '35', '72', '68',
                  ]
total = 0
for r in revenue_by_week:
    total += int(r)
print(total)
# I just realized that I suppose to fix this code down there but oh well...

1038


Use the skills you learned about lists, functions, and loops to get this data sorted out.

## Task 1: Fixing The Data

![](https://media.giphy.com/media/QsnQsvAkrkKGiHZuo8/giphy.gif)

Take a closer look at that error from the cell above, coming from running the `sum()` function. 

It's a TypeError. Python is basically saying that it doesn't know how to add (`+`) a number (`int`) and string (`str`) together. 

So now take a closer look at the `revenue_by_week` list. You can see that some of the numbers are there as numbers (without the surrounding `''` quotes), and some of them are strings (with surrounding `''` quotes).

While you could go into the list and manually remove all of the `'` marks, that sounds like a pain, and imagine if you had a lot more numbers than the three months of ASOS data. You can use your programming skills so that Python does the manual work for you! There's a built-in function, `int()`, that changes the argument given to it into a number. **Run the cell below** to see a demonstration on some various data types. 

<span style='background :#FFF59E'>**Note**: The last line in the cell will give an error since a `list` is not a number (the individual elements can be, though!)</span>

In [11]:
# Run this cell (it will give you an error)
print(int(105))   # integer
print(int(52.80)) # decimal (float)
print(int('97'))  # integer string
print(([105, '97'])) # list 
#I know we are not suppose to fix this but it bugs me (no puns intended) seeing an error in my code so I just delete the int() here.

105
52
97
[105, '97']


Notice that the `int()` function can accept values that are already numbers (chopping off any decimal part if present) or strings that depict integers (it will fail on decimal strings, however). But you should also notice that Python threw an error when we tried to give it a list.

In order to clean the data, we need to do that item by item. And how do we go item by item? Yes! We use a loop!

![](https://media.giphy.com/media/ryQDjtPSPfgeA/giphy.gif)

To complete this task, create a new list `cleaned_revenue` that has all of the data values in a numeric data type:
- Set up `cleaned_revenue` as an empty list.
- Use a `for` loop to loop over the elements of `revenue_by_week`.
  - For each element, convert it to an integer data type with the `int()` function
  - Append the converted value to the `cleaned_revenue` list.

Outside of your loop, `print` the completed `cleaned_revenue` and `print` the total sum of values.

In [74]:
# define the cleaned_revenue variable and set to zero
# you mean empty list not varaible right?
cleaned_revenue = []

# loop through data and convert to integers
for r in revenue_by_week:
    cleaned_revenue.append(int(r))

# print the cleaned_revenue variable
print (cleaned_revenue)

# print the total sum of values
total = 0
for c in cleaned_revenue:
    total += c
print(total)
# I tried sum(cleaned_revenue) but it did not work. The error said that 'int' object is callable. This is a weird error because when I tested my code in another coding tool, it works just fine. I hope you did not deduct my point by doing this way. 

[65, 77, 66, 74, 64, 82, 86, 72, 80, 96, 101, 35, 72, 68]
1038


<span style='background :#DDD5F3'>If done correctly, the value for the sum of `cleaned_revenue` should be **1038**</span>

## Task 2: Slicing and Dicing

![](https://media.giphy.com/media/vsuBq0HLMUw6nTqLWS/giphy.gif)

Your friend is so pleased with your work that they want you to do some additional calculations. 

- **What was the total amount made in each month?**
- **Which month had the highest average (weekly revenue?**

To answer these questions, you should know that the months are divided so that the first four weeks were June, the middle five weeks were July (5th-9th entries in list), and the last five weeks were August (10th-14th entries in list). 

Use slicing to get the relevant parts of the original revenue list, then use the `sum()` and `len()` functions to help you calculate the total and average for each month: the average will be the total revenue divided by the number of weeks. 

<span style='background :#FFF59E'>**Tip**: Be careful about how indexing works in Python! You might want to try printing the slices you pull out first to check that they're capturing the correct values, before trying to summarize them.</span>

In [81]:
# revenue by month 
jun = cleaned_revenue [:4]
jul = cleaned_revenue [4:9]
aug = cleaned_revenue [9:]

# my sum() function does not work as intended so I have to calculate the sum myself
tot_jun = 0
for m in jun:
    tot_jun += i
    
tot_jul = 0
for i in jul:
    tot_jul += i

tot_aug = 0
for i in aug:
    tot_aug += i

# calculate the averages 
avg_jun = tot_jun/len(jun)
avg_jul = tot_jul/len(jul)
avg_aug = tot_aug/len(aug)

# remember to use sum and len functions! :(



# print the total amount and average revenue for each month
print (f'The total amount of June is {tot_jun} millions pounds and the average revenue that month is {avg_jun} millions pounds')
print (f'The total amount of July is {tot_jul} millions pounds and the average revenue that month is {avg_jul} millions pounds')
print (f'The total amount of August is {tot_aug} millions pounds and the average revenue that month is {avg_aug} millions pounds')


The total amount of June is 272 millions pounds and the average revenue that month is 68.0 millions pounds
The total amount of July is 384 millions pounds and the average revenue that month is 76.8 millions pounds
The total amount of August is 372 millions pounds and the average revenue that month is 74.4 millions pounds


The month with the highest total and average avenue is July, with 384 millions pounds for total and 76.8 millions pounds for average. 

So, a little bit comment about my code above. As you have seen instead of using the sum() function, I went the long road instead and did 3 more for loops for my total. I have tried an hour debugging this but I can't, the 'int' object is callable for me is impossible to find. I have looked up Google to see why I have that mistake and apparently, that mistake only pops up if I use one of the build-in function to name a variable, which I apparently did. However, after knowing this error, I have changed all my "sum" variables into "total" right away, but the error is still there. I know that the problem is with the sum since max and min still work just fine. I even used ctrl+ f to search all the "sum" keywords in my code and I can't find it. What more frustrating is that my code works well in another coding tool (if you are curious, I ran my code on Replit and it works just fine)! So I gave up, I will just do the long way here. Forgive me for my inefficency in coding. I also tried not to repeat printing and calculating the same thing for each months but whenever I write a new code, it always ended up longer than intended (I used while loop and if/else statement to try to append them into a new list and so on). At the end, so much time wasted and I accomplished nothing. Please, I beg you to find the cause of the "callable" bug since that one drove me to the edge of insanity. Also, please forgive me for ranting like this and I hope I still get full credits for my work here still because of the correct results at least. 

<span style='background :#DDD5F3'>If done correctly the average values you should get are:
    
- June: 70.5
- July: 76.8
- August: 74.4
</span>